mmv1/templates/terraform/encoders/backend_service.go.erb

```diff
@@ -47,4 +47,43 @@ for _, backendRaw := range backends {
 	}
 }
 
+// This custom encoding helps prevent sending 0 for clientTtl, defaultTtl and
+// maxTtl in API calls to update these values  when unset in the provider
+// (doing so results in an API level error)
+c, cdnPolicyOk := d.GetOk("cdn_policy")
+
+// Only apply during updates
+if !cdnPolicyOk || obj["cdnPolicy"] == nil {
+	return obj, nil
+}
+
+currentCdnPolicies := c.([]interface{})
+
+// state does not contain cdnPolicy, so we can return early here as well
+if len(currentCdnPolicies) == 0 {
+	return obj, nil
+}
+
+futureCdnPolicy := obj["cdnPolicy"].(map[string]interface{})
+currentCdnPolicy := currentCdnPolicies[0].(map[string]interface{})
+
+cacheMode, ok := futureCdnPolicy["cache_mode"].(string)
+// Fallback to state if doesn't exist in object
+if !ok {
+	cacheMode = currentCdnPolicy["cache_mode"].(string)
+}
+
+switch cacheMode {
+case "USE_ORIGIN_HEADERS":
```

Comment:
Could we additionally confirm their values are `0`, not just that they're present? That'll avoid an invalid configuration passing because we're unintuitively shaping the request message.

Reply:
Without my changes to the backend service encoder the test fails with this following error:

```
Error: Error updating BackendService "projects/simpleclub-magic-modules-tests/global/backendServices/tf-test-894wd7irfa": googleapi: Error 400: Invalid value for field 'resource.cdnPolicy.defaultTtl': '3600'. default_ttl cannot be specified with USE_ORIGIN_HEADERS cache_mode., invalid
```

Meaning the TTL is 3600 and not 0.
If we would only change the values if they are 0, then the test would fail! I'm not 100% sure where the 3600 comes from, it is not mentioned in the test anywhere nor is it mentioned in the `BackendService.yaml` file.
There the TTL fields all have `default_from_api: true` set.

Reply:
@rileykarson I've looked at it again and tested changing the test to the following:

```go
switch cacheMode {
case "USE_ORIGIN_HEADERS":
	if clientTtl, ok := futureCdnPolicy["clientTtl"]; ok && clientTtl == 0 {
		delete(futureCdnPolicy, "clientTtl")
	}
	if defaultTll, ok := futureCdnPolicy["defaultTtl"]; ok && defaultTll == 0 {
		delete(futureCdnPolicy, "defaultTtl")
	}
	if maxTtl, ok := futureCdnPolicy["maxTtl"]; ok && maxTtl == 0 {
		delete(futureCdnPolicy, "maxTtl")
	}
}
```

Which then failed test with the following error:
```go
 Error: Error updating BackendService "projects/simpleclub-magic-modules-tests/global/backendServices/tf-test-ys68zkv7to": googleapi: Error 400: Invalid value for field 'resource.cdnPolicy.defaultTtl': '3600'. default_ttl cannot be specified with USE_ORIGIN_HEADERS cache_mode., invalid
        
          with google_compute_backend_service.foobar,
          on terraform_plugin_test.tf line 2, in resource "google_compute_backend_service" "foobar":
           2: resource "google_compute_backend_service" "foobar" {
        
--- FAIL: TestAccComputeBackendService_withCdnPolicy (147.38s)
```

-- 

My current theory of why this is happening is the following:
In the previous runs, the cache mode is NOT `USE_ORIGIN_HEADERS`, meaning the defaults for `clientTtl`, etc., would be fetched from the server and persisted in the state.

Then, because nothing changed and this is a PUT request, all existing data needs to be sent along, so also `clientTtl` with the default 3600 would be present in the `futureCdnPolicy`.
But now we don't want that to be present, so we should remove it even though previously it was 3600!

Reply:
After those further investigations, I believe we can't check if those values are null 🤔

Reply:
Right- yep, and the field are optional + computed so any old value will be preserved if the field is not explicitly specified. And the value can't be set to zero easily because it's optional + computed, so zero means null/unspecified some or all of the time too. OK- yeah, because of the update case we can't reliably check the value.